client/src/send_and_confirm_transactions_in_parallel.rs

```diff
@@ -0,0 +1,278 @@
+                    let current_block_height = current_block_height.load(Ordering::Relaxed);
+                    let transactions_to_verify: Vec<Signature> = transaction_map
+                        .iter()
+                        .filter(|x| {
+                            current_block_height < x.blockheight + MAX_RECENT_BLOCKHASHES as u64
+                        })
+                        .map(|x| *x.key())
+                        .collect();
+                    for signatures in
+                        transactions_to_verify.chunks(MAX_GET_SIGNATURE_STATUSES_QUERY_ITEMS)
+                    {
+                        if let Ok(result) = rpc_client.get_signature_statuses(signatures).await {
+                            let statuses = result.value;
+                            for (signature, status) in signatures.iter().zip(statuses.into_iter()) {
+                                if let Some(status) = status {
+                                    if status.satisfies_commitment(rpc_client.commitment()) {
+                                        if let Some((_, data)) = transaction_map.remove(signature) {
+                                            confirmed_transactions.fetch_add(1, Ordering::Relaxed);
+                                            let mut transaction_errors =
+                                                transaction_errors.lock().await;
+                                            transaction_errors[data.index] = status.err;
+                                        }
+                                    }
+                                }
+                            }
+                        }
+                    }
+                }
+                tokio::time::sleep(CONFIRMATION_REFRESH_RATE).await;
+            }
+        });
+    }
+
+    // transaction sender task
+    let total_transactions = messages.len();
+    let mut initial = true;
+    for expired_blockhash_retries in (0..5).rev() {
+        // only send messages which have not been confirmed
+        let nb_transaction = messages.len();
+        let messages: Vec<(usize, Message)> = if initial {
+            initial = false;
+            messages.iter().cloned().enumerate().collect()
+        } else {
+            // remove all the confirmed transactions
+            transaction_map
+                .iter()
+                .map(|x| (x.index, x.transaction.message.clone()))
+                .collect()
+        };
+
+        if messages.is_empty() {
+            break;
+        }
+
+        // clear the map so that we can start resending
+        transaction_map.clear();
+
+        for (index, message) in messages.iter() {
+            let mut transaction = Transaction::new_unsigned(message.clone());
+            let blockhash = *block_hash_rw.read().await;
+            transaction.try_sign(signers, blockhash)?;
+            let signautre = transaction.signatures[0];
+            if !tpu_client.send_transaction(&transaction).await {
```

Comment:
pretty sure there's a batch send version of this method

Reply:
Sending in batch is already used when we resend the transaction. I want to avoid it because here we send transaction over TPUClient and then over RPC client if it fails. 

The main thing here is the transaction is created and signed and send, making something like batches will complicate the logic and we will have something that exists already where the blockhash is expired while we are sending the transactions.